Fix collect_set_on_nested_type tests failed

thirtiseven opened the pull request:
Fixes #8716 

`collect_set` function currently does not support NaNs in struct[Array(Double)] or struct[Array(Float)] types, and NaNs are set to not equal to each others by it, see #6079. 

But the current IT will generate such data if we have bad luck, leading to test failure.

This PR prevents datagen from generating NaNs when testing `collect_set' on nested types.

<!--

Thank you for contributing to RAPIDS Accelerator for Apache Spark!

Here are some guidelines to help the review process go smoothly.

1. Please write a description in this text box of the changes that are being
   made.

2. Please ensure that you have written units tests for the changes made/features
   added.

3. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

4. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please label it with `[WIP]`.

5. If your pull request is ready to be reviewed without requiring additional
   work on top of it, then remove the `[WIP]` label (if present).

6. Once all work has been done and review has taken place please do not add
   features or make changes out of the scope of those requested by the reviewer
   (doing this just add delays as already reviewed code ends up having to be
   re-reviewed/it is hard to tell what is new etc!). Further, please avoid
   rebasing your branch during the review process, as this causes the context
   of any comments made by reviewers to be lost. If conflicts occur during
   review then they should be resolved by merging into the branch used for
   making the pull request.

Many thanks in advance for your cooperation!

-->

thirtiseven commented:
build

ttnghia commented:
What will happen if we feed in struct[Array(Double)] or struct[Array(Float)] types? Will it fallback to the CPU or something else?

jlowe reviewed (commented):
This doesn't look like the proper fix to resolve the test failure.  The test is failing because we don't match the CPU for some corner cases.  The fix isn't to ignore those corner cases in the test but rather correct the code so those corner cases can now pass the test.  If we cannot fix the corner cases in a timely manner, then we need to fallback to the CPU for those corner cases.  In this case, it would be falling back to the CPU if the nested types contain floats or doubles anywhere in the type tree.

thirtiseven commented:
> This doesn't look like the proper fix to resolve the test failure. The test is failing because we don't match the CPU for some corner cases.

Ok, thanks, I didn't notice that the hasNans config in the original PR has been removed, so it really is a bug now.

As [this comment](https://github.com/NVIDIA/spark-rapids/pull/6079#issuecomment-1199912967) said, usually when we use non-nested versions of floats and doubles, NaN values are considered unequal, but when collecting sets of nested array versions, NaN equality is considered on the CPU.

This incompatibility was controlled by `spark.rapids.sql.hasNans`, but we removed this config later.

> What will happen if we feed in struct[Array(Double)] or struct[Array(Float)] types? Will it fallback to the CPU or something else?

So now `collect_set` will produce different results without fallbacks.

I haven't found out why Spark behaves like this yet, the Spark code of `collect_set` looks just like add data into a HashSet then convert it to Array. But I guess it is because of some special handling of NaNs in Spark. If it is complicated to follow or needs cuDF support, I will fall back to cpu in this case and file another issue.

thirtiseven commented:
Some test results:
```python
>>> from pyspark.sql.functions import collect_set
>>> df1 = spark.createDataFrame([(1.0,), (float("nan"),), (float("nan"),)], ["value"])
>>> df1.agg(collect_set("value")).show(truncate=False)
23/07/25 18:07:49 WARN GpuOverrides:
*Exec <ProjectExec> will run on GPU
  *Expression <Alias> cast(collect_set(value)#4 as string) AS collect_set(value)#7 will run on GPU
    *Expression <Cast> cast(collect_set(value)#4 as string) will run on GPU
  *Exec <ObjectHashAggregateExec> will run on GPU
    *Expression <AggregateExpression> collect_set(value#0, 0, 0) will run on GPU
      *Expression <CollectSet> collect_set(value#0, 0, 0) will run on GPU
    *Expression <Alias> collect_set(value#0, 0, 0)#2 AS collect_set(value)#4 will run on GPU
    *Exec <ShuffleExchangeExec> will run on GPU
      *Partitioning <SinglePartition$> will run on GPU
      *Exec <ObjectHashAggregateExec> will run on GPU. The data type of following expressions will be converted in GPU runtime: buf#10: Converted BinaryType to ArrayType(DoubleType,false)
        *Expression <AggregateExpression> partial_collect_set(value#0, 0, 0) will run on GPU
          *Expression <CollectSet> collect_set(value#0, 0, 0) will run on GPU
        ! <RDDScanExec> cannot run on GPU because GPU does not currently support the operator class org.apache.spark.sql.execution.RDDScanExec
          @Expression <AttributeReference> value#0 could run on GPU

+------------------+
|collect_set(value)|
+------------------+
|[NaN, NaN, 1.0]   |
+------------------+

>>> df2 = spark.createDataFrame([([1.0, float("nan")],), ([1.0, float("nan")],), ([1.0, 2.0],)], ["value"])
>>> df2.agg(collect_set("value")).show(truncate=False)
23/07/25 18:07:57 WARN GpuOverrides:
*Exec <ProjectExec> will run on GPU
  *Expression <Alias> cast(collect_set(value)#28 as string) AS collect_set(value)#31 will run on GPU
    *Expression <Cast> cast(collect_set(value)#28 as string) will run on GPU
  *Exec <ObjectHashAggregateExec> will run on GPU
    *Expression <AggregateExpression> collect_set(value#24, 0, 0) will run on GPU
      *Expression <CollectSet> collect_set(value#24, 0, 0) will run on GPU
    *Expression <Alias> collect_set(value#24, 0, 0)#26 AS collect_set(value)#28 will run on GPU
    *Exec <ShuffleExchangeExec> will run on GPU
      *Partitioning <SinglePartition$> will run on GPU
      *Exec <ObjectHashAggregateExec> will run on GPU. The data type of following expressions will be converted in GPU runtime: buf#34: Converted BinaryType to ArrayType(ArrayType(DoubleType,true),false)
        *Expression <AggregateExpression> partial_collect_set(value#24, 0, 0) will run on GPU
          *Expression <CollectSet> collect_set(value#24, 0, 0) will run on GPU
        ! <RDDScanExec> cannot run on GPU because GPU does not currently support the operator class org.apache.spark.sql.execution.RDDScanExec
          @Expression <AttributeReference> value#24 could run on GPU

+------------------------------------+
|collect_set(value)                  |
+------------------------------------+
|[[1.0, NaN], [1.0, 2.0], [1.0, NaN]]|
+------------------------------------+

>>> spark.conf.set("spark.rapids.sql.enabled", "false")
>>> df1.agg(collect_set("value")).show(truncate=False)
+------------------+
|collect_set(value)|
+------------------+
|[1.0, NaN, NaN]   |
+------------------+

>>> df2.agg(collect_set("value")).show(truncate=False)
+------------------------+
|collect_set(value)      |
+------------------------+
|[[1.0, 2.0], [1.0, NaN]]|
+------------------------+
>>>
```

thirtiseven commented:
Updated code to fallback to cpu if there is double/float in nested type, and filed issue #8808

revans2 commented on integration_tests/src/main/python/hash_aggregate_test.py:
```diff
@@ -688,12 +696,20 @@ def test_hash_groupby_collect_set(data_gen):
+@allow_non_gpu('ObjectHashAggregateExec', 'ShuffleExchangeExec', 'CollectSet')
+@pytest.mark.parametrize('data_gen', _gen_data_for_collect_set_op_floats, ids=idfn)
+def test_hash_groupby_collect_set_fallback_on_nested_floats(data_gen):
+    assert_gpu_and_cpu_are_equal_collect(
```
please use `assert_gpu_fallback_collect` instead.  It will help to verify that we actually did fall back on what we expect.

thirtiseven replied:
Ok, done.

revans2 commented on integration_tests/src/main/python/hash_aggregate_test.py:
```diff
@@ -731,12 +747,18 @@ def test_hash_reduction_collect_set(data_gen):
+@allow_non_gpu('ObjectHashAggregateExec', 'ShuffleExchangeExec', 'CollectSet')
+@pytest.mark.parametrize('data_gen', _gen_data_for_collect_set_op_floats, ids=idfn)
+def test_hash_reduction_collect_set_fallback_on_nested_floats(data_gen):
+    assert_gpu_and_cpu_are_equal_collect(
```
Same comment here. if the test is for a fallback we should use the fallback verification API.

revans2 commented on integration_tests/src/main/python/window_function_test.py:
```diff
@@ -1120,10 +1120,10 @@ def do_it(spark):
     ('c_struct_array_2', RepeatSeqGen(StructGen([
-        ['c0', struct_array_gen_no_nans], ['c1', int_gen]]), length=14)),
-    ('c_array_struct', RepeatSeqGen(ArrayGen(all_basic_struct_gen_no_nan), length=15)),
+        ['c0', struct_array_gen_no_floats], ['c1', int_gen]]), length=14)),
```
nit: Can we verify that we fallback for window operations too?

thirtiseven replied:
Added.

revans2 commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala:
```diff
@@ -3264,18 +3264,13 @@ object GpuOverrides extends Logging {
-              }
-            case ArrayType(et, _) => et.isInstanceOf[ArrayType] || et.isInstanceOf[StructType]
-            case _ => false
+        override def tagAggForGpu(): Unit = {
```
Why not just remove TypeSig.FLOAT and TypeSig.DOUBLE from the supported types?

```
 expr[CollectSet](
      "Collect a set of unique elements, not supported in reduction",
      // https://github.com/NVIDIA/spark-rapids/issues/8808 fall back in case there might be NaNs
      ExprChecks.fullAgg(
        TypeSig.ARRAY.nested(TypeSig.commonCudfTypes + TypeSig.DECIMAL_128 +
            TypeSig.NULL + TypeSig.STRUCT + TypeSig.ARRAY - TypeSig.FLOAT - TypeSig.DOUBLE),
        TypeSig.ARRAY.nested(TypeSig.all),
        Seq(ParamCheck("input",
          (TypeSig.commonCudfTypes + TypeSig.DECIMAL_128 +
              TypeSig.NULL +
              TypeSig.STRUCT +
              TypeSig.ARRAY - TypeSig.FLOAT - TypeSig.DOUBLE).nested(),
          TypeSig.all))),
...
```

This should update the docs and the make it more clear that we do not support these types.

thirtiseven commented:
build

jlowe commented on integration_tests/src/main/python/hash_aggregate_test.py:
```diff
@@ -603,23 +603,31 @@ def test_hash_pivot_groupby_duplicates_fallback(data_gen):
     ('a', RepeatSeqGen(LongGen(), length=20)),
     ('b', value_gen)] for value_gen in _repeat_agg_column_for_collect_set_op]
 
+_gen_data_for_collect_set_op_floats = [[
+    ('a', RepeatSeqGen(LongGen(), length=20)),
+    ('b', RepeatSeqGen(struct_array_gen, length=15))]]
```
This only checks arrays containing floats but not structs containing floats.  We removed floats from structs up above, those ones we removed should be appearing here.

thirtiseven replied:
done.

jlowe commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala:
```diff
@@ -3264,18 +3264,13 @@ object GpuOverrides extends Logging {
           TypeSig.all))),
```
There should be a PS note generated for struct and array noting that if they contain floating point types within they will not be supported.

thirtiseven replied:
done.

thirtiseven commented:
build
